### ISSUE TITLE
Need class for containers of elements that have their own keys

### ISSUE THREAD

**Issue by AshleyYakeley** [snoyberg/mono-traversable#129]
Currently, there's the `IsMap` class:
```
class (MonoTraversable map, SetContainer map) => IsMap map where
    type MapValue map
    lookup :: ContainerKey map -> map -> Maybe (MapValue map)
    insertMap :: ContainerKey map -> MapValue map -> map -> map
    -- etc.
```
In most (all?) instances of `IsMap`, we have `Element map ~ (ContainerKey map, MapValue map)`, that is, there's some `MapValue` type that is "the element minus the key".

However, for some containers, there isn't a suitable `MapValue` type because the key cannot easily be removed from the `Element` type. Instead, we'd want a class that looks more like this:

```
class (MonoTraversable map, SetContainer map) => KeyContainer map where
    elementKey :: Element map -> ContainerKey map
    lookupElement :: ContainerKey map -> map -> Maybe (Element map)
    insertElement :: Element map -> map -> map
    -- etc.
```

**Comment by snoyberg** [comment]
Can you give an example of one of these other container types you're
talking about?

**Comment by AshleyYakeley** [comment]
I don't know if such types already exist, but:

1. A hash-addressable store, where `Element` is `ByteString` and `ContainerKey` is a SHA-256 hash.
1. An in-memory typed SQL table, where `Element` is a row, and `ContainerKey` is the primary key.

**Comment by snoyberg** [comment]
I'm sorry, but at this level of abstraction I'm having a really hard time understanding the request here.

**Comment by AshleyYakeley** [comment]
Fair enough. Given that I don't actually know of any types in existing Haskell packages that would implement this and not `IsMap`, it might not be worth it.
